src/spartan/mod.rs

```diff
@@ -225,3 +204,124 @@ fn compute_eval_table_sparse<E: Engine>(
 
   (A_evals, B_evals, C_evals)
 }
+
+#[cfg(all(test, not(target_arch = "wasm32")))]
+mod tests {
+  use super::*;
+  use crate::provider::PallasEngine;
+  use crate::r1cs::util::{FWrap, GWrap};
+  use pasta_curves::pallas::Point as PallasPoint;
+  use pasta_curves::Fq as Scalar;
+  use proptest::collection::vec;
+  use proptest::prelude::*;
+
+  impl<E: Engine> PolyEvalWitness<E> {
+    fn alt_batch(p_vec: &[&Vec<E::Scalar>], s: &E::Scalar) -> Self {
+      p_vec
+        .iter()
+        .skip(1)
+        .for_each(|p| assert_eq!(p.len(), p_vec[0].len()));
+
+      let powers_of_s = powers::<E>(s, p_vec.len());
+
+      let p = zip_with!(par_iter, (p_vec, powers_of_s), |v, weight| {
+        // compute the weighted sum for each vector
+        v.iter().map(|&x| x * *weight).collect::<Vec<E::Scalar>>()
+      })
+      .reduce(
+        || vec![E::Scalar::ZERO; p_vec[0].len()],
+        |acc, v| {
+          // perform vector addition to combine the weighted vectors
+          acc.into_iter().zip_eq(v).map(|(x, y)| x + y).collect()
+        },
+      );
+
+      Self { p }
+    }
+  }
+
+  impl<E: Engine> PolyEvalInstance<E> {
```

Comment:
Now with generic `PolyEvalInstance` we can derive various implementations of it, right? Do we use that already (or have plans to create and use) and if so where concretely (more exactly - do we have / need alternative implementation which is different comparing to the previous [reference](https://github.com/lurk-lab/Nova/blob/af601ed0513a2df6321a2d88293ca3260e71ef84/src/spartan/mod.rs#L87) one implemented at Solidity side of things)?

Reply:
`PolyEvalInstance` has been generic since the beginning, and the type itself has not changed:
https://github.com/microsoft/Nova/blob/37871d04ba04771bf19c51e5b336e45fdb86c7f7/src/spartan/mod.rs#L117-L121
I'm not sure I understand the question?

Reply:
Right, the question is actually - does new PolyEvalInstanse preserve compatibility with older one and where alt_batch is planning to be used (since as far as I see it is absent in upstream)